### ISSUE TITLE
linter: Add allowMemoization option to react/no-unstable-nested-components

### ISSUE THREAD

**Issue by Hideman42** [oxc-project/oxc#23812]
Hi,

First of all, yes I'm aware that:
  1. This isn't an option in the [official rule](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md)
  2. There is a warning saying this is still a bad practice and should be avoided

The will is to allow some cases while disallowing bad cases.
For instance this is really useful in translation where you dont want to put it outside but keep the code and understanding at the same place.
```tsx
export function FileUpload(/* ... */) {
  /* ... */
  const id = useId();
  retrun (
    <div>
      <p className="file-input-text pe-none">
        <FormattedMessage
          id="fileUpload.dndDescription"
          values={useMemo(
            // allowing memoization would allow this usage
            () => ({
              label: (txt): ReactElement => (
                <label htmlFor={id} tabIndex={0} className="link link-info" role="button">
                  {txt}
                </label>
              ),
            }),
            [id],
          )}
        />
      </p>
      <input
        className="d-none"
        type="file"
        name="fileInput"
        id={id}
        accept="text/csv"
        onChange={useCallback(
          evt => {
            onFileSelected(evt.target.files?.item(0));
            evt.target.value = '';
          },
          [onFileSelected],
        )}
      />
    </div>
  );
}
```

This would be very great to support this, today I have too many usages with translation that I have to silent the rule completely where I'm sure it could detect truly bad usage.

Please let me know what you think about this.

Thanks in advance 🙏
